contrib/inventory/cloudforms.py

```diff
@@ -0,0 +1,351 @@
+            print "======================================================================="
+            print "======================================================================="
+            print json.dumps(results, indent=2)
+            print "======================================================================="
+            print "======================================================================="
+            print "======================================================================="
+
+        return results
+
+    def _get_hosts(self):
+        """
+        Get all hosts by paging through the results
+        """
+        limit = self.cloudforms_limit
+
+        page = 0
+        last_page = False
+
+        results = []
+
+        while not last_page:
+            offset = page * limit
+            ret = self._get_json("%s/api/vms?offset=%s&limit=%s&expand=resources,tags&attributes=id,ipaddresses,name,power_state" % (self.cloudforms_url, offset, limit))
+            results += ret['resources']
+            if ret['subcount'] < limit:
+                last_page = True
+            page += 1
+
+        return results
+
+    def _get_host_by_id(self, hid):
+        return self._get_json("%s/api/vms/%s" % (self.cloudforms_url, hid))
+
+    def _resolve_host(self, host):
+        """
+        Resolve host details.
+        """
+        host_id = host['id']
+        ret = self._get_host_by_id(host_id)
+
+        if self.cloudforms_purge_actions:
+            del ret['actions']
+
+        return ret
+
+    def update_cache(self):
+        """
+        Make calls to cloudforms and save the output in a cache
+        """
+        self.groups = dict()
+        self.hosts = dict()
+
+        if self.args.debug:
+            print "Updating cache..."
+
+        for host in self._get_hosts():
+            if host['power_state'] != 'on':
+                if self.args.debug:
+                    print "Skipping %s because power_state = %s" % (host['name'], host['power_state'])
+                continue
+
+            if self.cloudforms_prefer_ip_address and host.has_key('ipaddresses') and host['ipaddresses'][0]:
+                dns_name = host['ipaddresses'][0]
+            else:
+                dns_name = host['name']
+
+            if self.args.debug:
+                print "Host named [%s] will use [%s] as dns_name" % (host['name'], dns_name)
+
+            # Create ansible groups for hostgroup, location and organization
+            if host.has_key('tags'):
+                for group in host['tags']:
+                    safe_key = self.to_safe(group['name'])
+                    self.push(self.inventory, safe_key, dns_name)
+
+                    if self.args.debug:
+                        print "Found tag [%s] for host which will be mapped to [%s]" % (group['name'], safe_key)
+
+            full_host = self._resolve_host(host)
+            if host.has_key('ipaddresses'):
```

Comment:
Can you use `if 'ipaddresses' in host` instead?  We should use this where ever you have `has_key()`